### ISSUE TITLE
Cache files are never closed on upload

### ISSUE THREAD

**Issue by scarfacedeb** [refile/refile#414]
I've noticed that after I've attached the file to the AR record and saved that record, there're always leftover opened file descriptors from cache.

It's not a big deal in small quantities, but it could potentially lead to `Too many open files` errors. 

I've got that error in my application due to my own mistakes.
But it forced me to investigate other sources of never-closed files.

I added a test to demonstrate the issue: https://github.com/scarfacedeb/refile/commit/51cc2b2d540c1f92d99c1815f6f783ac8b383d39

Right now it fails.

**Comment by scarfacedeb** [comment]
Moreover, I was able to fix the test by always closing the `uploadable` in `FileSystem#upload`: https://github.com/scarfacedeb/refile/commit/78c5837aa534e9295f545e9d2365933a27e2a381

All other tests are green, but I'm not sure that it's the right thing to do in this case.

**Comment by scarfacedeb** [comment]
Also it seems to me that in jruby `IO.copy_stream(src, dest)` always creates file descriptor inside ObjectSpace, which lead to 2 extra never-closed files after `record.save`.  (one for cache and one for store).

To reproduce it you can remove `.reject(&:closed?).select { |f| f.path.include? "cache" }` from the counters in the test.

In MRI, it still works as before, but in jruby it fails due to 2 extra files.

I was able to fix that issue by refactoring `upload` like this:
```ruby
        dest = ::File.open path(id), "wb"
        IO.copy_stream(uploadable, dest)
        ...
    ensure
        uploadable.close
        dest.close
```

But if I recall correctly, it's a bad idea due to performance.

**Comment by jnicklas** [comment]
Super interesting! I seem to recall that there was some reason we don't close the file on upload, but I can't remember at all why that would be, and I can't think of any good reason. Maybe something like leaving that responsibility to the caller? That seems error prone though.

Eventually file descriptors should be released when the GC collects the file objects and finalizes them, from what I understand, so I'm not sure how worrying this really is?

As for `copy_stream` in JRuby, I doubt that there's any (not super hacky) way of working around that, might be worth reporting to the JRuby project though.

**Comment by scarfacedeb** [comment]
@jnicklas I conducted couple of experiments regarding file descriptors release by GC.

I was using modified test from above:

```ruby

    it "closes opened cache files" do
      before_files_count = ObjectSpace.each_object(File).reject(&:closed?).count

      3000.times do
        post = klass.new
        post.document = Refile::FileDouble.new("hello")
        post.save
      end

      after_files_count = ObjectSpace.each_object(File).reject(&:closed?).count

      GC.start # for  MRI

      require "jruby"
      JRuby.gc # for jruby

      expect(after_files_count).to eq(before_files_count)
    end
```

GC in MRI was more predictable and it left less than 50 opened files in ObjectSpace by the end of the test.

GC in jruby was less thorough and sometimes it left ~2000 opened files files in ObjectSpace after 3000 times.

Surprisingly, when I increased the number of times to 5000, it produced much better results: 100-900 files.

Overall, it seems to me that relying on GC (in jruby) to release those file descriptors could work, but it's less predictable, than releasing them manually. (Could be relevant: https://github.com/jruby/jruby/issues/221)

And unfortunately, closing just the `uploadable` won't solve the problem completely due to `IO.copy_stream` bug (?) in jruby.

But I'm no expert on this topic.
